Package MTCFeatures: https://pvankranenburg.github.io/MTCFeatures/

Pour récupérer les MIDI ou pdf pour que ce soit plus visuel (en particulier, c'est le dataset MTC-FS-INST (900Mb)): https://www.liederenbank.nl/mtc

In [3]:
# décommenter pour installer le package une fois pour toutes
!pip install MTCFeatures

In [1]:
import MTCFeatures
from MTCFeatures import MTCFeatureLoader
import pandas as pd
import numpy as np

In [2]:
# Run once
MTCFeatures.downloadData(dest='user')

[██████████████████████████████████████████████████]MTC-ANN-2.0.1_sequences-1.1.jsonl.gz?download=1 -> /home/marwane/.local/share/MTCFeatures/MTC-ANN-2.0.1_sequences-1.1.jsonl.gz
[██████████████████████████████████████████████████]MTC-FS-INST-2.0_sequences-1.1.jsonl.gz?download=1 -> /home/marwane/.local/share/MTCFeatures/MTC-FS-INST-2.0_sequences-1.1.jsonl.gz
[██████████████████████████████████████████████████]essen_sequences-1.1.jsonl.gz?download=1 -> /home/marwane/.local/share/MTCFeatures/essen_sequences-1.1.jsonl.gz


In [3]:
fl = MTCFeatureLoader('MTC-FS-INST-2.0')
seqs = fl.sequences()

In [5]:
type(seqs)

generator

In [6]:
item = seqs.__next__()
#print(item)
#print(item.keys())
#print(item['features'].keys())

In [7]:
%%time
phrase_data = []
for ii, x in enumerate(seqs):
    phrase_data.append({
        'id': x['id'],
        **x['features']
    })

CPU times: user 12 s, sys: 1.44 s, total: 13.5 s
Wall time: 13.5 s


In [8]:
df_phrase = pd.DataFrame(phrase_data)

for i in range(5):
    df_pitch = pd.DataFrame(df_phrase["pitch"].iloc[i])
    df_mode = pd.DataFrame(df_phrase["mode"].iloc[i])
    print(df_pitch.nunique(), df_mode.nunique())


0    13
dtype: int64 0    1
dtype: int64
0    11
dtype: int64 0    1
dtype: int64
0    10
dtype: int64 0    1
dtype: int64
0    9
dtype: int64 0    1
dtype: int64
0    7
dtype: int64 0    1
dtype: int64


In [9]:
df_phrase.columns

Index(['id', 'scaledegree', 'scaledegreespecifier', 'tonic', 'mode',
       'metriccontour', 'imaweight', 'pitch40', 'midipitch', 'diatonicpitch',
       'diatonicinterval', 'chromaticinterval', 'pitchproximity',
       'pitchreversal', 'nextisrest', 'restduration_frac', 'duration',
       'duration_frac', 'duration_fullname', 'onsettick', 'beatfraction',
       'phrasepos', 'phrase_ix', 'phrase_end', 'songpos', 'beatinsong',
       'beatinphrase', 'beatinphrase_end', 'IOI_frac', 'IOI', 'IOR',
       'imacontour', 'pitch', 'contour3', 'contour5', 'beatstrength',
       'beat_str', 'beat_fraction_str', 'beat', 'timesignature',
       'gpr2a_Frankland', 'gpr2b_Frankland', 'gpr3a_Frankland',
       'gpr3d_Frankland', 'gpr_Frankland_sum', 'lbdm_spitch', 'lbdm_sioi',
       'lbdm_srest', 'lbdm_rpitch', 'lbdm_rioi', 'lbdm_rrest',
       'lbdm_boundarystrength', 'durationcontour', 'IOR_frac', 'lyrics',
       'noncontentword', 'wordend', 'phoneme', 'rhymes', 'rhymescontentwords',
       'word

In [10]:
df_phrase['beatfraction']


0        [3/4, 1/4, 1, 1, 1, 1, 1/2, 1/4, 1/4, 1/2, 1/2...
1        [1/2, 1/2, 1/2, 3/2, 1/2, 3/2, 1/2, 2, 1, 3/2,...
2        [1/2, 1/2, 1/2, 1, 1, 1/2, 1/2, 1, 1/2, 1/2, 3...
3        [1/2, 1/4, 1/4, 1/2, 1/4, 1/4, 1/2, 1/2, 1, 1/...
4        [1, 1/3, 1/3, 1/3, 1, 1, 2/3, 1/3, 1/2, 1/6, 1...
                               ...                        
18103    [1/2, 1/2, 1/2, 1/2, 1/2, 1/2, 1, 1/2, 1/2, 1/...
18104    [1/3, 1/3, 1/3, 2/3, 1/3, 1/3, 1/3, 2/3, 1/3, ...
18105    [1/2, 1/2, 1/2, 1/2, 1/2, 1/4, 1/4, 1/2, 1/2, ...
18106    [3/8, 1/8, 3/8, 1/8, 3/8, 1/8, 3/8, 1/8, 3/8, ...
18107    [1/3, 1/3, 1/3, 1/3, 2/3, 1/3, 1/3, 1/3, 1/3, ...
Name: beatfraction, Length: 18108, dtype: object

In [11]:
df_phrase['phrase_end']

0        [False, False, False, False, False, False, Fal...
1        [False, False, False, False, False, False, Fal...
2        [False, False, False, False, False, False, Fal...
3        [False, False, False, False, False, False, Fal...
4        [False, False, False, False, False, False, Fal...
                               ...                        
18103    [False, False, False, False, False, False, Fal...
18104    [False, False, False, False, False, False, Fal...
18105    [False, False, False, False, False, False, Fal...
18106    [False, False, False, False, False, False, Fal...
18107    [False, False, False, False, False, False, Fal...
Name: phrase_end, Length: 18108, dtype: object

In [42]:
from fractions import Fraction
sequence_length = 5
selected_features = ['midipitch','diatonicinterval', 'duration', 'scaledegree', 'diatonicpitch', 'songpos', 'nextisrest']
#, 'beatstrength', 'duration', 'scaledegree', 'diatonicpitch', 'songpos', 'IOI', 'nextisrest'

In [44]:
'''X = []  # matrice 3D pour les sous-séquences
y = []  # Liste pour les labels

for index, row in df_phrase.iterrows():
    # Récupération des séquences pour chaque attribut
    sequences = {feature: row[feature] for feature in selected_features}
    phrase_end = row['phrase_end']  # Récupération des labels de fin de phrase
    # Création des sous-séquences
    for i in range(len(phrase_end) - sequence_length + 1):
        # Vérifie si on peut extraire une sous-séquence valide
        subsequence = {feature: sequences[feature][i:i + sequence_length] for feature in selected_features}
        subsequence_label = phrase_end[i + sequence_length - 1]  # Dernier élément comme label
        
        # Ajouter la sous-séquence à X et le label à y
        X.append([subsequence[feature] for feature in selected_features])
        y.append(subsequence_label)

# Conversion en matrices numpy
X = np.array(X)  # Shape: (N, L, F)
y = np.array(y)'''

"X = []  # matrice 3D pour les sous-séquences\ny = []  # Liste pour les labels\n\nfor index, row in df_phrase.iterrows():\n    # Récupération des séquences pour chaque attribut\n    sequences = {feature: row[feature] for feature in selected_features}\n    phrase_end = row['phrase_end']  # Récupération des labels de fin de phrase\n    # Création des sous-séquences\n    for i in range(len(phrase_end) - sequence_length + 1):\n        # Vérifie si on peut extraire une sous-séquence valide\n        subsequence = {feature: sequences[feature][i:i + sequence_length] for feature in selected_features}\n        subsequence_label = phrase_end[i + sequence_length - 1]  # Dernier élément comme label\n        \n        # Ajouter la sous-séquence à X et le label à y\n        X.append([subsequence[feature] for feature in selected_features])\n        y.append(subsequence_label)\n\n# Conversion en matrices numpy\nX = np.array(X)  # Shape: (N, L, F)\ny = np.array(y)"

In [46]:
#X[0]

In [55]:
X2 = []  # Liste pour les sous-séquences
y2 = []  # Liste pour les labels

for index, row in df_phrase.iterrows():
    # Récupération des séquences pour chaque attribut
    sequences = {feature: row[feature] for feature in selected_features}
    phrase_end = row['phrase_end']  # Récupération des labels de fin de phrase

    if "diatonicinterval" in sequences:
        sequences["diatonicinterval"][0] = -1
    if "nextisrest" in sequences:
        sequences["nextisrest"][-1] = True
        sequences["nextisrest"] = [int(value) for value in sequences["nextisrest"]]
    # Création des sous-séquences
    for i in range(len(phrase_end) - sequence_length + 1):
        # Extraction de la sous-séquence
        subsequence = {feature: sequences[feature][i:i + sequence_length] for feature in selected_features}
        subsequence_phrase_end = phrase_end[i:i + sequence_length]  # Sous-séquence des labels phrase_end
        
        # Vérifie si la sous-séquence contient au moins une fin de phrase (phrase_end = True)
        if any(subsequence_phrase_end):  
            # Ajouter la sous-séquence à X
            X2.append([subsequence[feature] for feature in selected_features])
            # Le label est la valeur de phrase_end à la dernière position de la sous-séquence
            y2.append(subsequence_phrase_end[-1])

# Conversion en matrices numpy
X2 = np.array(X2)  # Shape: (N, L, F)
y2 = np.array(y2)  # Shape: (N,)
X2.shape

(483639, 7, 5)

In [56]:
X2[1]

array([[ 6.90000000e+01,  6.90000000e+01,  6.90000000e+01,
         7.40000000e+01,  7.20000000e+01],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         3.00000000e+00, -1.00000000e+00],
       [ 2.50000000e-01,  5.00000000e-01,  5.00000000e-01,
         2.00000000e+00,  5.00000000e-01],
       [ 3.00000000e+00,  3.00000000e+00,  3.00000000e+00,
         6.00000000e+00,  5.00000000e+00],
       [ 3.00000000e+01,  3.00000000e+01,  3.00000000e+01,
         3.30000000e+01,  3.20000000e+01],
       [ 5.27522936e-02,  5.50458716e-02,  5.96330275e-02,
         6.42201835e-02,  8.25688073e-02],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  0.00000000e+00]])

In [59]:
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import SGDClassifier, LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

linr = LinearRegression()
#poly = PolynomialFeatures(degree=3)
svc = SVC()
lr = LogisticRegression(max_iter=5000)
knn = KNeighborsClassifier(7)
sdg = SGDClassifier()

dt = DecisionTreeClassifier()
mlp = MLPClassifier(max_iter=5000)

methods = [linr, lr, knn, sdg, dt, mlp]

In [61]:
from sklearn.model_selection import cross_val_score
acc = {}
std = {}
X_flattened = X2.reshape(X2.shape[0], -1)
for m in methods:
    acc[m] = []
    std[m] = []
    r = cross_val_score(m, X_flattened, y2)
    acc[m].append(np.mean(r))
    std[m].append(np.std(r))
    print("{:>25} {:1.3f} ({:1.3f})".format(m.__class__.__name__, np.mean(r), np.std(r)))


         LinearRegression 0.482 (0.030)
       LogisticRegression 0.899 (0.008)
     KNeighborsClassifier 0.850 (0.007)
            SGDClassifier 0.892 (0.010)
   DecisionTreeClassifier 0.889 (0.009)
            MLPClassifier 0.909 (0.009)


# Cross val differente méthode

# midipitch 
LinearRegression 0.072 (0.006)

LogisticRegression 0.760 (0.002)

KNeighborsClassifier 0.781 (0.008)

SGDClassifier 0.700 (0.120)

DecisionTreeClassifier 0.800 (0.007)

MLPClassifier 0.778 (0.004)

# midipitch duration 
LinearRegression 0.396 (0.010)

LogisticRegression 0.875 (0.003)

KNeighborsClassifier 0.854 (0.007)

DecisionTreeClassifier 0.866 (0.005)

MLPClassifier 0.887 (0.004)

# ['midipitch', 'duration', 'songpos']
LinearRegression 0.392 (0.010)

LogisticRegression 0.873 (0.003)

KNeighborsClassifier 0.865 (0.005)

DecisionTreeClassifier 0.871 (0.006)

MLPClassifier 0.885 (0.002)

# ['midipitch', 'duration', 'songpos','diatonicpitch']
LinearRegression 0.397 (0.009

LogisticRegression 0.875 (0.003)

KNeighborsClassifier 0.856 (0.006)

DecisionTreeClassifier 0.871 (0.006)

MLPClassifier 0.887 (0.004)

# ['midipitch','diatonicinterval', 'duration', 'scaledegree', 'diatonicpitch', 'songpos', 'nextisrest']
Transformation faites pour nextisrest, et diatonicpitch : 

LinearRegression 0.482 (0.030)

LogisticRegression 0.899 (0.008)

KNeighborsClassifier 0.850 (0.007)

SGDClassifier 0.892 (0.010)

DecisionTreeClassifier 0.889 (0.009)

MLPClassifier 0.909 (0.009)

# Attibuts non fonctionnel hors DT
## midipitch, ioi
DecisionTreeClassifier 0.887 (0.007)
## + nextisrest 
DecisionTreeClassifier 0.890 (0.008)
## + nextisrest + beatstrength
DecisionTreeClassifier 0.903 (0.006)
## + beatstrength
DecisionTreeClassifier 0.901 (0.006)
